File: kotlin/hydroponics-host/hydroponics-host/src/main/kotlin/io/greenglass/hydroponics/application/models/implementation/ImplementationModel.kt
package io.greenglass.hydroponics.application.models.implementation

import org.neo4j.graphdb.*

import io.greenglass.host.application.neo4j.GraphNode
import io.greenglass.hydroponics.application.models.implementation.ImplementationNodeModel.Graph.IMPLEMENTATION_NODE
import io.greenglass.hydroponics.application.models.system.HydroponicsSystemModel

class ImplementationModel(
    val implementationId : String,
    val systemId : String,
    val name : String,
    val nodes : List<ImplementationNodeModel>

) : GraphNode {

    constructor(node : Node) : this(
        implementationId = node.getProperty("implementationId") as String,
        systemId = node.getSingleRelationship(HydroponicsSystemModel.SYSTEM,Direction.OUTGOING)
            .endNode
            .getProperty("systemId") as String,
        name = node.getProperty("name") as String,
        nodes = node.getRelationships(IMPLEMENTATION_NODE).map { r -> ImplementationNodeModel(r.endNode) }
    )

    override fun toNode(tx: Transaction): Node {
        val node = tx.createNode(implementationLabel)
        val systemNode = checkNotNull(tx.findNode(HydroponicsSystemModel.systemLabel,"systemId", systemId ))
        node.createRelationshipTo(systemNode, HydroponicsSystemModel.SYSTEM)
        node.setProperty("implementationId", implementationId)
        node.setProperty("name", name)
        nodes.forEach { n -> node.createRelationshipTo(n.toNode(tx), IMPLEMENTATION_NODE)}
        return node
    }

    companion object Graph {
        val implementationLabel : Label = Label.label("Implementation")
        val IMPLEMENTATION : RelationshipType = RelationshipType.withName("IMPLEMENTATION")
    }
}

File: flutter/hydroponics_framework/lib/datamodels/implementation/implementation_model.dart
import 'implementation_node_model.dart';

class ImplementationModel {
  late String implementationId;
  late String name;
  late String systemId;
  late List<ImplementationNodeModel> nodes;

  ImplementationModel(
      this.implementationId,
      this.name,
      this.systemId,
      this.nodes);

  ImplementationModel.fromJson(Map<String,dynamic> json) {
    implementationId = json["implementationId"];
    name = json["name"];
    systemId = json["systemId"];
    nodes = List<Map<String,dynamic>>.from(json["nodes"])
        .map((n) => ImplementationNodeModel.fromJson(n))
        .toList();
  }
}